Shipping rates UI changes: wp.data

ecgan opened the pull request:
### Changes proposed in this Pull Request:

- Part of a PR refactor from the huge PR https://github.com/woocommerce/google-listings-and-ads/pull/1218.
- Depends on changes from the following PRs: 
  - https://github.com/woocommerce/google-listings-and-ads/pull/1202
  - https://github.com/woocommerce/google-listings-and-ads/pull/1249

This PR contains wp.data changes to make things work with the new shipping rates API changes. The `upsertShippingRates` and `deleteShippingRates` are removed and replaced with `saveShippingRates` that is able to handle both within the function.

In subsequent PRs, shipping rates in the UI will be stored in the form state, and `saveShippingRates` will be called with the complete shipping rates array from the form state.

### Detailed test instructions:

This is wp.data API changes, there isn't a UI way to test it in this PR branch.

You can test this PR using browser console like this:

```js
wp.data.resolveSelect( 'wc/gla' ).getShippingRates().then( ( rates ) => {
  // Mutate `rates` here if want to test rate updating or deletion
  wp.data.dispatch( 'wc/gla' ).saveShippingRates( rates );
} )
```

Alternatively, you can check out PR #1218 which contains the same wp.data changes, and test the Setup MC or Edit Free Listings flows, and you would be able to observe the API calls from there.

<!--
Optional.
Enter a summary of all changes in this Pull Request, which will be added to the changelog if accepted.
Each line should start with `(Fix|Add|Tweak|Update) - `, for example:
> Fix - I took care of something that wasn't working.
> Add - I added something new that's pretty cool.
> Tweak - I made a small change.
> Update - I made big changes to something that wasn't broken.

Leave the "Changelog entry" header in place completely empty, without any summary if no changelog entry is needed.
If you remove the "Changelog entry" header, the title of Pull Request will be used as the changelog entry.
-->
### Changelog entry

>

ecgan committed:
Change shipping rates code in wp-data.

Upsert and delete shipping rates are not really needed, we only need saveShippingRates.

ecgan committed:
Change upsertShippingRates to saveShippingRates in EditFreeCampaign.

ecgan committed:
Remove obsolete code comments.

ecgan committed:
Fix notice should be success, not error, when saving is successful.

jorgemd24 commented on js/src/data/actions.js:
```diff
@@ -72,77 +73,80 @@ export function* fetchShippingRates() {
+
+			if ( ! found ) {
+				acc.push( cur.id );
+			}
```
❓ 

What do you think if we do something like the code below so it is more descriptive and easier to test?

```suggestion

const isInNewShippingRates = ( item,  newShippingRates = [] ) => {
  return newShippingRates.some( ( el ) =>
  isSameShippingRate( el, item )
  );
}

.......

const deleteIds = oldShippingRates.filter((item) => ! isInNewShippingRates( item, newShippingRates ) ).map( item => item.id);


```
We could also add some tests for this part of the code.

ecgan replied:
I did some changes (see [`18ee6a6` (#1262)](https://github.com/woocommerce/google-listings-and-ads/pull/1262/commits/18ee6a62df8acb20d7f064cdd72a7c5bc2d506b1)) based on your suggestion, and I do think they are much easier to read now. I have also added tests for it. Thanks for the idea! 🙂  👍

jorgemd24 commented on js/src/data/actions.js:
```diff
@@ -72,77 +73,80 @@ export function* fetchShippingRates() {
-export function* upsertShippingRates( shippingRate ) {
-	const { countryCodes, currency, rate } = shippingRate;
-
+export function* saveShippingRates( newShippingRates ) {
```
❓ 

What do you think if the `saveShippingRates` is compound by 2 functions for example:

```
function* saveShippingRates{
   //Here we check if there are rates that can be deleted and we do the request.
   const deleteIds = maybeDeleteRates( oldShippingRates, newShippingRates);
  
   const upsertedShippingRates = upsertedRates(newShippingRates)
   .....
}
```

Like this: 
- It will be easier to test each part. 
- Each function will have a smaller responsibility.
- Each function will have a specific logic.

ecgan replied:
I tried out your suggestion in [`567342d` (#1262)](https://github.com/woocommerce/google-listings-and-ads/pull/1262/commits/567342dfe9988229441b6b912c4a7c28f9349d7a), I have a bit of mix feelings about it, specifically I don't quite like the `maybeDeleteShippingRates` with "maybe" in it, but I guess I can live with it. 😄

jorgemd24 replied:
I have left a comment related to that. See https://github.com/woocommerce/google-listings-and-ads/pull/1262#discussion_r818004417. However,  depending on https://github.com/woocommerce/google-listings-and-ads/pull/1262#discussion_r817832807 the full implementation could be different.

jorgemd24 commented on js/src/utils/isSameShippingRate.js:
```diff
@@ -0,0 +1,20 @@
+const isSameShippingRate = ( a, b ) => {
+	return (
+		( a.id !== undefined && a.id === b.id ) ||
+		( a.country === b.country && a.method === b.method )
```
❔ 

Could you explain me a bit more why is needed this part: `( a.country === b.country && a.method === b.method )` and why can we not rely only on the ids?

ecgan replied:
So we are trying to compare two shipping rates, and one of them would be coming from the UI form state. Imagine a scenario like this: users load the page with existing shipping rates (the shipping rates would have the ids, since they are loaded from API). Those shipping rates would be put into form state, and the UI would depend on the form state. When users add a new shipping rate, it would be added into the form state, and it has **no ID**, since it has not been saved yet. So to be safe and defensive, we also check on the country and method combination.

Also, on the API side, we also checked on country and method combination to determine whether there are existing shipping rate: 

https://github.com/woocommerce/google-listings-and-ads/blob/4bfc1c723b7a80414b40bb578178a765ebaeed7b/src/API/Site/Controllers/MerchantCenter/ShippingRateController.php#L185

jorgemd24 replied:
Thanks, @ecgan for the explanation. 😃

jorgemd24 reviewed (commented):
Thanks, @ecgan, it is looking good. I have tested using PR #1218 and all is working as expected 👍 

I have left some comments about the coding and also it would be nice if we add tests for some of the functions.

eason9487 commented on js/src/data/actions.js:
```diff
@@ -72,77 +73,80 @@ export function* fetchShippingRates() {
+		 * Compare the new shipping rates and the old ones from the store
+		 * to find out the old ones to be deleted.
+		 */
+		const oldShippingRates = yield select( STORE_KEY, 'getShippingRates' );
```
❔ Would you consider passing `oldShippingRates` through the function caller instead of getting it inside this function? The reason is basically the same as https://github.com/woocommerce/google-listings-and-ads/pull/1262#discussion_r814721197.

ecgan replied:
> Would you consider passing `oldShippingRates` through the function caller instead of getting it inside this function?

I suppose you mean changing `saveShippingRates` from the current one:

```
function* saveShippingRates( newShippingRates )
```

to something like this below? 

```
function* saveShippingRates( oldShippingRates, newShippingRates )
```

At first I thought about that (before I found out about `yield select` in actions), but I much prefer the current one, because it doesn't require the consumer to get the old shipping rate just to call `saveShippingRates`. For example, in later Setup MC PR https://github.com/woocommerce/google-listings-and-ads/pull/1267/files#diff-90f9623f145c55b1acc58938118d5552c1d7a9d43909acabd630f6558a249e40R20, we have a change in `useAutoClearShippingEffect` where we call `saveShippingRate( [] )` with empty array to clear shipping rates, we don't need to use `useShippingRates()` to get the old shipping rates, the code looks succinct and nice in my opinion. 😄 

The way I think about `saveShippingRates` is similar to the current `saveSettings`, we only need the new values in the function parameter, we don't need the old ones.

eason9487 replied:
It is more flexible to design action with as much closer single responsibility as possible since actions are the elementary data control interfaces on frontend side.

For example, the current design of `saveShippingRates` fully overwrites all rates, which is like a RESTful **PUT** API call. When one or more shipping rates need to be deleted, the current state must be obtained from wp-data store, and filter out the rates to be deleted before calling `saveShippingRates`; when some shipping rates need to be updated, the current state still must be obtained from wp-data store, use that state to update the required ones, and then call `saveShippingRates` along with all rates that include those ones don't need updates.

That is, except it looks much cleaner on the caller side when clearing all shipping rates. But in other places, the caller always already has the rates state obtained from wp-data store. And with calling of `useShippingRates`, clearing all rates doesn't add much complexity.

Moreover, they also cause redundant API calls and database operations on the server-side internally. Especially when many countries are selected, the auto-save feature causes more redundant calls and loads to server.
https://github.com/woocommerce/google-listings-and-ads/blob/0cca8d9e92d620d21688dedf3bc2053c6f70cb07/src/API/Site/Controllers/MerchantCenter/ShippingRateBatchController.php#L54-L58

Moving `oldShippingRates` out of `saveShippingRates` could preserve a certain degree of flexibility for easier detail optimization in the future. And it should be easier to write tests for `saveShippingRates` because it would not need to mock `select`.

---

If go further into abstraction design, I would prefer to keep the `upsertShippingRates` and `deleteShippingRates` actions as they were exported originally. The `saveShippingRates` would be better suited as a custom hook, since it actually handles business logics that are closer to UI-wise, but now those business logics are highly coupled into data layer with data processing logic. In this way,
- the caller side could still not need to obtain the current shipping rates state if caller doesn't have it.
- the data layer would have more flexibility.
- it's more stateless so that writing tests should be easier.
- the interfaces of `upsert*` and `delete*` actions would be more clear than `saveShippingRates( oldShippingRates, newShippingRates )` when partially updating/deleting.
- it would also be easier to handle exceptions of API call in the future.

ecgan replied:
> I would prefer to keep the `upsertShippingRates` and `deleteShippingRates` actions as they were exported originally. _The `saveShippingRates` would be better suited as a custom hook_, since it actually handles business logics that are closer to UI-wise

Hey I think that is a good idea! 😃 👍  So to summarize, I think this is what we are thinking:

1. In wp.data, we don't need `saveShippingRates`, we want `upsertShippingRates` and `deleteShippingRates` actions.
2. We come up with a custom hook like the following, which consumes the above `upsertShippingRates` and `deleteShippingRates` actions:

```js
// an example:
const useSaveShippingRate = () => {
	const saveShippingRate = useCallback( () => {
		// do the deleteShippingRates and upsertShippingRates here.
	}, [])

	return saveShippingRate
}
```

3. Change the code in Edit Free Listings and Setup MC to use the above hook.

This approach also addresses @jorgemd24's comment in https://github.com/woocommerce/google-listings-and-ads/pull/1262#discussion_r814721197, to have better function composition.

Given that the current `saveShippingRate` action is being used in Edit Free Listings PR https://github.com/woocommerce/google-listings-and-ads/pull/1266 and Setup MC PR https://github.com/woocommerce/google-listings-and-ads/pull/1267, I will work on that change as a separate PR later based on https://github.com/woocommerce/google-listings-and-ads/pull/1267.

ecgan replied:
> At first I thought about that (before I found out about `yield select` in actions)

Side note for posterity: I found put about the `yield select` in actions from here: https://unfoldingneurons.com/2020/wordpress-data-store-properties-action-creator-generators

ecgan replied:
> Moreover, they also cause redundant API calls and database operations on the server-side internally. Especially when many countries are selected, the auto-save feature causes more redundant calls and loads to server.

Ahh yes, I recall that I thought about this before. We can address this within the `saveShippingRates (newShippingRates)` function (action or custom hook):

1. Delete the old shipping rates, like what we already have.
2. For the new shipping rates, we loop through them and compare with the shipping rate object found from wp.data store. We should do a deep comparison here, to cater for `shippingRate.options.free_shipping_threshold` etc. If they are not equal (i.e. they have been changed), we add the new shipping rate to an array pending to be upserted.
3. With the array of shipping rates to be upserted, we call upsertShippingRates, and get back the result.
4. Merge the result with the unchanged shipping rates and put them into wp.data store.

With this approach, we will only call upsertShippingRate for those shipping rates that are newly created or have been changed, and the consumer would still call the same API `saveShippingRates (newShippingRates)`, which is nice. I didn't implement this because I thought it was somewhat of an early optimization, however we can certainly do that later.

eason9487 replied:
> 1\. [...]
> 2\. [...]
> 3\. Change the code in Edit Free Listings and Setup MC to use the above hook.

Yes, the `useSaveShippingRate` could still do the same thing inside the hook, and further have the improvement of https://github.com/woocommerce/google-listings-and-ads/pull/1262#discussion_r817858184. And if some other callers already know what to upsert/delete, they could call `upsertShippingRates`/`deleteShippingRates` directly.

> Given that the current `saveShippingRate` action is being used in Edit Free Listings PR #1266 and Setup MC PR #1267, I will work on that change as a separate PR later based on #1267.

> With this approach, we will only call upsertShippingRate for those shipping rates that are newly created or have been changed, and the consumer would still call the same API `saveShippingRates (newShippingRates)`, which is nice.

Agree with these. 👍  Thanks for considering these options!

ecgan replied:
> Given that the current `saveShippingRate` action is being used in Edit Free Listings PR #1266 and Setup MC PR #1267, I will work on that change as a separate PR later based on #1267.

FYI, I have created issue https://github.com/woocommerce/google-listings-and-ads/issues/1287 to keep track of this.

eason9487 commented on js/src/data/actions.js:
```diff
@@ -72,77 +73,80 @@ export function* fetchShippingRates() {
 			},
 		} );
 
+		const upsertedShippingRates = data.success?.map( ( el ) => el.rate );
```
🚧  The type of `rate.rate` here is inconsistent with the above.

https://github.com/woocommerce/google-listings-and-ads/blob/bf1bfbdc480f4845831c08bbee1c0d3ead5d2467/js/src/data/actions.js#L56

ecgan replied:
Good catch, fixed in [`aa33008` (#1262)](https://github.com/woocommerce/google-listings-and-ads/pull/1262/commits/aa33008bf1c3a4cb29dbfaabe9fab571ade628b6).

eason9487 commented on js/src/data/actions.js:
```diff
@@ -72,77 +73,80 @@ export function* fetchShippingRates() {
 			},
 		} );
 
+		const upsertedShippingRates = data.success?.map( ( el ) => el.rate );
```
❓ or 🚧 
Is there any case the API will respond without `success` property? If so, `data.success?.map()` will result in `undefined` and break the page.

ecgan replied:
Good thinking! I looked into the PHP code (in `feature/refactor-shipping-rates-api` branch) and it seems like `success` array should always be there, so we should be safe here. 😄 

https://github.com/woocommerce/google-listings-and-ads/blob/4bfc1c723b7a80414b40bb578178a765ebaeed7b/src/API/Site/Controllers/MerchantCenter/ShippingRateBatchController.php#L48-L74

eason9487 replied:
💅 

> I looked into the PHP code (in `feature/refactor-shipping-rates-api` branch) and it seems like `success` array should always be there, so we should be safe here.

Thanks for checking it. I'd suggest the `map` could be called by `.map` to avoid doubts about whether any edge cases may have occurred here.

ecgan replied:
Gotcha, done in [`b01c91b` (#1262)](https://github.com/woocommerce/google-listings-and-ads/pull/1262/commits/b01c91ba31d64c36c403b30a41e830b29ff17d72). 🙂

eason9487 reviewed (commented):
Since the structure of `ShippingRate` was changed, could you also update this JSDoc accordingly?

https://github.com/woocommerce/google-listings-and-ads/blob/bf1bfbdc480f4845831c08bbee1c0d3ead5d2467/js/src/data/actions.js#L34-L41

ecgan committed:
Update JSDoc for shipping rate.

ecgan committed:
Convert the upsertedShippingRates' rate into number.

This is to make saveShippingRates consistent with fetchShippingRates.

ecgan commented:
@eason9487 ,

> Since the structure of `ShippingRate` was changed, could you also update this JSDoc accordingly?

Done in [`c2deb39` (#1262)](https://github.com/woocommerce/google-listings-and-ads/pull/1262/commits/c2deb398821d91d82e736007ad87f6338e28bc53).

ecgan committed:
Code refactor with isInShippingRates.

ecgan committed:
Refactor saveShippingRates into smaller functions.

ecgan committed:
Update code comments.

ecgan committed:
Remove unneeded optional chaining.

jorgemd24 commented on js/src/data/actions.js:
```diff
@@ -71,78 +75,91 @@ export function* fetchShippingRates() {
+	}
+}
+
+function* upsertShippingRates( shippingRates ) {
```
❓

Is it not convenient to put these functions `upsertShippingRates` &  `maybeDeleteShippingRates` out of this file (action.js) ? The reason why I think it would be convenient is that they are not real actions, for example, it does not return a `type` and a `new state` for the reducer.

ecgan replied:
Thanks, I will change these in a new PR, in relation to my comment in https://github.com/woocommerce/google-listings-and-ads/pull/1262#discussion_r817832807. 🙂 👍

jorgemd24 commented on js/src/data/actions.js:
```diff
@@ -71,78 +75,91 @@ export function* fetchShippingRates() {
- */
-export function* upsertShippingRates( shippingRate ) {
-	const { countryCodes, currency, rate } = shippingRate;
+function* maybeDeleteShippingRates( newShippingRates, oldShippingRates ) {
```
❔

I was just looking the code again and I think this function should be just `deleteShippingRates( deleteIds = [])` and the logic to get the deleteIds could go in something like `const deleteIds = getDeleteIds( newRates, oldRates)`.
Like this deleteShippingRates is agnostic about how to get the deleteIds and it could be reused easily and the testing should be only on `getDeleteIds` that is the part doing the logic part.

What do you think?

ecgan replied:
> the logic to get the deleteIds could go in something like `const deleteIds = getDeleteIds( newRates, oldRates)`.
> Like this deleteShippingRates is agnostic about how to get the deleteIds and it could be reused easily and the testing should be only on `getDeleteIds` that is the part doing the logic part.

That sounds good. I'll look into that in my new PR later, in relation to my comment in https://github.com/woocommerce/google-listings-and-ads/pull/1262#discussion_r817832807. 🙂 👍

jorgemd24 reviewed (commented):
Thanks @ecgan for these changes 👍 

 I would like to test the actual changes using this PR but It is a bit hard as it depends on others PRs.

At the moment I am getting the error `TypeError: upsertShippingRates is not a function` in the following line:

https://github.com/woocommerce/google-listings-and-ads/blob/0d649834ef6196f2153781ebb53587451537371f/js/src/setup-mc/setup-stepper/setup-free-listings/shipping-rate/shipping-rate-setup/countries-price-input/edit-rate-button/edit-rate-modal/index.js#L58

I think it is because this file has been updated in another PR. 

I left a couple of comment about the actions functions.

eason9487 commented on js/src/utils/isInShippingRates.test.js:
```diff
@@ -0,0 +1,174 @@
+			method: 'flat_rate',
+			currency: 'USD',
+			rate: 25,
+			options: [],
```
Note: Related to the initial values mentioned in https://github.com/woocommerce/google-listings-and-ads/pull/1266#discussion_r816551941

Or maybe we could remove all `options` from this test since they don't affect the test target.

ecgan replied:
Gotcha, I have removed them in [`fcd905b` (#1262)](https://github.com/woocommerce/google-listings-and-ads/pull/1262/commits/fcd905ba48a0478ebaf8b3d79c3f1fca40e22645).

eason9487 commented:
Hey @jorgemd24 

> I would like to test the actual changes using this PR but It is a bit hard as it depends on others PRs.

FYI, the way I tested this PR was using the Console tab of browser DevTool to run
```js
wp.data.resolveSelect( 'wc/gla' ).getShippingRates().then( ( rates ) => {
  // Mutate `rates` here if want to test rate updating or deletion
  wp.data.dispatch( 'wc/gla' ).saveShippingRates( rates );
} )
```

eason9487 reviewed (approved):
Thanks for the updates! The https://github.com/woocommerce/google-listings-and-ads/pull/1262#discussion_r818289055 could be adjusted in separate PR. The current implementation is ready to proceed with merging.

ecgan commented:
> FYI, the way I tested this PR was using the Console tab of browser DevTool to run

TIL! I have updated the PR description with that info. Thanks @eason9487 ! 👍

ecgan committed:
Remove options from isInShippingRates test.

`options` is supposed to be an object, but when there are no options, the API returned it as an array, which can be confusing. Removing the options array here for now.
